// @port-labs/plugins-sdk/dist/pluginHostMessaging.js
const THEME_STYLE_ELEMENT_ID = 'port-plugin-theme';
const ACCEPTED_ORIGIN_REGEX = /^(?:http:\/\/localhost(?::\d+)?|http:\/\/127\.0\.0\.1(?::\d+)?|https:\/\/localhost(?::\d+)?|https:\/\/(?:[a-z0-9-]+\.)*getport\.io(?::\d+)?|https:\/\/(?:[a-z0-9-]+\.)*port\.io(?::\d+)?)$/;
const initialState = {
    params: {},
    page: undefined,
    user: undefined,
    entity: undefined,
    theme: undefined,
    token: null,
    portApiBaseUrl: null,
};
function parseTheme(value) {
    if (value === null || typeof value !== 'object') {
        return undefined;
    }
    const record = value;
    if (typeof record.mode !== 'string' || typeof record.css !== 'string') {
        return undefined;
    }
    return { mode: record.mode, css: record.css };
}
let state = initialState;
const listeners = new Set();
let listening = false;
let requestedPortToken = false;
let portOpenInParentBridgeInstalled = false;
function postOpenLinkToParent(url) {
    window.parent.postMessage({ type: 'OPEN_URL', data: { url } }, '*');
}
function createFakePopupWindow() {
    return {
        closed: false,
        close: () => { },
        focus: () => { },
        blur: () => { },
        postMessage: () => { },
    };
}
function onDocumentClickCaptureOpenInParent(event) {
    if (window.parent === window)
        return;
    const el = event.target;
    if (!(el instanceof Element))
        return;
    const anchor = el.closest('a[href]');
    if (!(anchor instanceof HTMLAnchorElement))
        return;
    const href = anchor.href;
    if (!href || href.startsWith('javascript:'))
        return;
    event.preventDefault();
    event.stopImmediatePropagation();
    postOpenLinkToParent(href);
}
export function installPortOpenInParentBridge() {
    if (typeof window === 'undefined' || portOpenInParentBridgeInstalled)
        return;
    portOpenInParentBridgeInstalled = true;
    function patchedOpen(url) {
        const urlString = url == null ? '' : typeof url === 'string' ? url : url.href;
        postOpenLinkToParent(urlString);
        return createFakePopupWindow();
    }
    window.open = patchedOpen;
    if (typeof document !== 'undefined') {
        document.addEventListener('click', onDocumentClickCaptureOpenInParent, true);
    }
}
function notify() {
    for (const listener of listeners) {
        listener();
    }
}
function setState(partial) {
    state = { ...state, ...partial };
    notify();
}
function onMessage(event) {
    const isValidOrigin = ACCEPTED_ORIGIN_REGEX.test(event.origin);
    if (window.parent !== window && event.source !== window.parent)
        return;
    if (!isValidOrigin)
        return;
    if (event.data?.type === 'PORT_TOKEN') {
        setState({ token: event.data.token ?? null });
        return;
    }
    if (event.data?.type === 'PLUGIN_DATA') {
        setState({
            params: (event.data.params ?? {}),
            page: (event.data.page ?? {}),
            user: (event.data.user ?? undefined),
            entity: (event.data.entity ?? undefined),
            theme: parseTheme(event.data.theme),
            portApiBaseUrl: event.data.baseUrl ?? null,
        });
    }
}
/**
 * Ensures a single `window` message listener is registered and requests a token from the host
 * when embedded in an iframe (once per page load). Safe to call multiple times; helps when the
 * host mounts the plugin twice (e.g. React Strict Mode) without missing the token.
 */
export function initPortPluginMessaging() {
    if (typeof window === 'undefined') {
        return;
    }
    installPortOpenInParentBridge();
    if (listening) {
        return;
    }
    listening = true;
    window.addEventListener('message', onMessage);
    if (window.parent !== window && !requestedPortToken) {
        requestedPortToken = true;
        window.parent.postMessage({ type: 'REQUEST_PORT_TOKEN' }, '*');
    }
}
export function subscribe(listener) {
    initPortPluginMessaging();
    listeners.add(listener);
    return () => {
        listeners.delete(listener);
    };
}
export function getSnapshot() {
    initPortPluginMessaging();
    return state;
}
const serverSnapshot = { ...initialState };
export function getServerSnapshot() {
    return serverSnapshot;
}
export function getToken() {
    initPortPluginMessaging();
    return state.token;
}
export function getUser() {
    initPortPluginMessaging();
    return state.user;
}
export function getEntity() {
    initPortPluginMessaging();
    return state.entity;
}
export function getPage() {
    initPortPluginMessaging();
    return state.page;
}
export function getParams() {
    initPortPluginMessaging();
    return state.params;
}
export function getPortApiBaseUrl() {
    initPortPluginMessaging();
    return state.portApiBaseUrl;
}
export function getTheme() {
    initPortPluginMessaging();
    return state.theme;
}
/**
 * Injects or updates a `<style id="port-plugin-theme">` in `document.head` using the latest
 * `theme.css` from the host. Removes that element when there is no theme or no `css`. No-op on
 * the server.
 */
export function applyThemeCss() {
    initPortPluginMessaging();
    if (typeof document === 'undefined') {
        return;
    }
    const css = state.theme?.css;
    const existing = document.getElementById(THEME_STYLE_ELEMENT_ID);
    if (!css) {
        existing?.remove();
        return;
    }
    const styleEl = existing ?? document.createElement('style');
    styleEl.id = THEME_STYLE_ELEMENT_ID;
    styleEl.textContent = css;
    if (!existing) {
        document.head.appendChild(styleEl);
    }
}
installPortOpenInParentBridge();
